### Part 1. Common code

#### Creating a list of dictionaies with overall data fields from the original data file

In [2]:
import re

In [12]:
pattern = re.compile(r'\d+')

In [13]:
# Creating a lists of fields and data rows 
data_rows = []

with open('StudentsPerformance.csv') as f:

    for line in f:
        info = line.split(',')
        if info[0] == '"gender"':
            data_fields = [field[1:-1] for field in info]
            continue
        else:
            data_rows.append(info)

In [14]:
# Creating the list of dictionaries
data_dict_list = []

columns_cnt = len(data_fields)

for row in data_rows:
    data_row_dict = dict(zip(data_fields, row))
    for field in data_fields:
        search_result = pattern.search(data_row_dict[field])
        if search_result:
            data_row_dict[field] = int(search_result[0])
        else:
            data_row_dict[field] = data_row_dict[field][1:-1]
    data_dict_list.append(data_row_dict)

In [12]:
def check_matches(pattern, strings):
    pattern = re.compile(pattern)
    results = [pattern.match(string)!=None for string in strings]
    return dict(zip(strings, results))

In [15]:
data_dict_list[:1]

[{'gender': 'female',
  'race/ethnicity': 'group B',
  'parental level of education': "bachelor's degree",
  'lunch': 'standard',
  'test preparation course': 'none',
  'math score': 72,
  'reading score': 72,
  'writing score"': 74}]

### Task 1.

Дана строковая переменная с именем string, содержащая одно слово на неизвестном языке. Напишите код, с помощью которого можно перевернуть эту переменную задом наперёд, т.е. получить слово, которое получается при чтении исходного слова справа налево.

Напишите ваш код в одну строку без пробелов. Не присваивайте результат работы кода никакой переменной

### Common code to check ourselves

In [16]:
string = 'string'

### Solution:

In [17]:
string[::-1]

'gnirts'

### Task 2.

Отметьте истинные утверждения:
- 'Hello' in 'Hello, Dolly!'.lower()
- 'ARE'.lower() in 'how are you?'.upper()
- 'your' in 'WhAt Is YoUr NaMe?'.lower()
- 'apples'.upper() in 'I like APPLES'.upper()

### Solution:

In [18]:
'Hello' in 'Hello, Dolly!'.lower()

False

In [20]:
'ARE'.lower() in 'how are you?'.upper()

False

In [21]:
'your' in 'WhAt Is YoUr NaMe?'.lower()

True

In [22]:
'apples'.upper() in 'I like APPLES'.upper()

True

### Task 3.

Отметьте варианты кода, после выполнения которых в переменной Х будет храниться вещественное число. Если иное не предлагается в варианте ответа, то считаем, что X ='55,66'.

- X = 3.14
- float(X.replace(',', '.'))
- X = int(X.replace(',', '.'))
- X = float(X.replace(',', '.'))

### Solution

In [26]:
X = 3.14
type(X) == float

True

##### Conclusion: this way is right. Great!!!

In [27]:
X = '55,66'

In [28]:
type(float(X.replace(',', '.'))) == float

True

##### Conclusion: it seems to be right way, but there is no assignment to X variable. So this way is wrong.

In [29]:
type(int(X.replace(',', '.'))) == float

ValueError: invalid literal for int() with base 10: '55.66'

##### Conclusion: this way is totally wrong. Attempt to convert '55.66' to int raises the 'ValueError' runtime exception.

In [32]:
X = '55,66'

In [33]:
X = float(X.replace(',', '.'))
type(X) == float

True

##### Conclusion: this way is right. Great!!!

### Task 4.

Отметьте символ, который не используется в качестве символа-джокера (метасимвола) при работе с регулярными выражениями.

- Восклицательный знак - ! верно
- Вопросительный знак - ?
- Точка - .
- Звёздочка - *
- Плюс - +

### Solution

##### Conclusion: '!' is not a regular expressions joker symbol.

### Task 5.

Отметьте строки, в которых есть совпадения со следующим шаблоном:

r'[АЕИ][а-я]+\s\-\s[а-я]+'

- Африка - моя любовь!
- Арбуз полезен для здоровья.
- А - первая буква в алфавите.
- Ах - это вам не ох!

### Solution

In [18]:
pattern = r'[АЕИ][а-я]+\s\-\s[а-я]+'

strings = [
    'Африка - моя любовь!',
    'Арбуз полезен для здоровья.',
    'А - первая буква в алфавите.',
    'Ах - это вам не ох!'
]

check_matches(pattern, strings)

{'Африка - моя любовь!': True,
 'Арбуз полезен для здоровья.': False,
 'А - первая буква в алфавите.': False,
 'Ах - это вам не ох!': True}

##### Conclusion: 1'st and 4'th ways are right.

### Task 6.

Отметьте строки, в которых есть совпадения со следующим шаблоном:

r'\W\d*\w+\s!'

- !13ff !
- ~sss !
- *123456789abcdefgh !
- =13x !

### Solution

In [19]:
pattern = r'\W\d*\w+\s!'

strings = [
    '!13ff !',
    '~sss !',
    '*123456789abcdefgh !',
    '=13x !'
]

check_matches(pattern, strings)

{'!13ff !': True, '~sss !': True, '*123456789abcdefgh !': True, '=13x !': True}

##### Conclusion: all ways are right.

### Task 7.

Какое значение будет храниться в переменной secret после выполнения кода, представленного ниже?
- None
- [2, 2, 4]
- ['2', '2', '4']
- Пустой список верно
- Пустой словарь

### Solution

In [20]:
import re
pattern = re.compile('\d\d\d')
secret = pattern.findall('2 x 2 = 4') 

In [21]:
secret

[]

##### Conclusion: the result is empty list.

### Task 8.

Какой средний балл, полученный мальчиками на экзамене по чтению? Ответ округлите до трёх цифр после запятой.

### Solution

In [36]:
males_cnt = sum(data_row['gender'] == 'male'
                for data_row in data_dict_list)
males_reading_score_sum = sum(data_row['reading score'] for data_row in data_dict_list if data_row['gender'] == 'male')
males_reading_score_avg = males_reading_score_sum / males_cnt
print(f'Средний бал, полученный мальчиками на экзамене по чтению: {males_reading_score_avg:.3f}')

Средний бал, полученный мальчиками на экзамене по чтению: 65.473


#### Conclusion: average reading tests score for males is 65.473

### Task 9.

Какой средний балл на экзамене по чтению набрали ученики, набравшие максимальный балл на экзамене по математике? Ответ округлите до трёх цифр после запятой.

In [38]:
max_math_score = 100
top_math_score_males_cnt = 0
top_math_score_males_reading_score_sum = 0
for row in data_dict_list:
    math_score = row['math score']
    if math_score == max_math_score:
        top_math_score_males_cnt += 1
        top_math_score_males_reading_score_sum += row['reading score']
        max_math_score = math_score

top_math_score_males_reading_score_avg = top_math_score_males_reading_score_sum / top_math_score_males_cnt

print(f'Средний балл теста на чтение мальчиков, сдавших математику на максимальные баллы: {top_math_score_males_reading_score_avg:.3f}')

Средний балл теста на чтение мальчиков, сдавших математику на максимальные баллы: 97.857


##### Conclusion: Average reading test score for males, who gain maximum scores on math tests is about 97.857. So, well mathematicians are strong readers as well.

In [34]:
data_dict_list[:1]

[{'gender': 'female',
  'race/ethnicity': 'group B',
  'parental level of education': "bachelor's degree",
  'lunch': 'standard',
  'test preparation course': 'none',
  'math score': 72,
  'reading score': 72,
  'writing score"': 74}]

### Task 10.

Какой средний балл на экзамене по письму набрали ученики, которые плохо победали перед экзаменом (lunch = free/reduced)? Ответ округлите до двух цифр после запятой.

In [41]:
poor_lunch_males_cnt = 0
poor_lunch_males_writing_scores_sum = 0

for row in data_dict_list:
    if row['lunch'] == 'free/reduced':
        poor_lunch_males_cnt += 1
        poor_lunch_males_writing_scores_sum += row['writing score"']
        
poor_lunch_males_writing_scores_avg = poor_lunch_males_writing_scores_sum / poor_lunch_males_cnt

print(f'Cредний балл на экзамене по письму среди учеников, которые плохо победали перед экзаменом: {poor_lunch_males_writing_scores_avg:.2f}')

Cредний балл на экзамене по письму среди учеников, которые плохо победали перед экзаменом: 63.02
